### ISSUE TITLE
Citation parser fails for statutes with letters in the section number

### ISSUE THREAD

**Issue by jmesserschmidt1** [freelawproject/eyecite#146]
U.S. code statutes with letters in them appear to be unrecognized. So "18 U.S.C. § 1028" and "18 U.S.C. § 1028(a)" are parsed, but "18 U.S.C. § 1028A" is not. I've tried some variations, but seems to be consistent.

**Comment by mlissner** [comment]
Thanks for sending this along. I think this would be pretty easy to fix, but our code parsers aren't particularly advanced compared to our opinions parsers. 

Do you want to take a stab at it?

**Comment by jmesserschmidt1** [comment]
> Thanks for sending this along. I think this would be pretty easy to fix, but our code parsers aren't particularly advanced compared to our opinions parsers.
> 
> Do you want to take a stab at it?

Sure. Not super familiar with the code, but suspect might need a variation on the law_section regex similar to the one that exists for page or volume, like [here](https://github.com/freelawproject/reporters-db/blob/22a42ce1bc8311631a7370a9efd5dc366d119c38/reporters_db/data/regexes.json#L68). This comes up with CFR cites as well (e.g., 17 CFR § 240.10b-5 is currently parsed as 17 CFR 240).  So something like `(?P<section>\\d+(?:[\\-.:]\\d+){,3})[a-zA-Z]{0,4})` and `(?P<section>\\d+(?:[\\-.:]\\d+){,3})[a-zA-Z]{0,4})`

**Comment by mlissner** [comment]
I don't know that part of the code very well either, but if you want to do a PR with tests that fixes this, I think we'd probably merge it (and release a new version, if desired).

**Comment by linhtpham09** [comment]
Hello, I am interested in picking up this ticket if it is still available?

**Comment by mightymikesapp** [comment]
Picking this up alongside #331 and #335, which fix different halves of it. I checked out both branches and ran the same strings against main, #331, and #335:

| citation | main | #331 | #335 |
|---|---|---|---|
| 18 U.S.C. § 1028A | `§` | FullLawCitation ✓ | ShortLawCitation `§` (uppercase not matched) |
| 17 U.S.C. § 106A | `§` | FullLawCitation ✓ | ShortLawCitation `§` |
| 15 U.S.C. § 78j(b) | `§` | `§` | ShortLawCitation `§ 78j(b)` |
| 42 U.S.C. § 2000e-2 | `§` | `§` | ShortLawCitation `§ 2000e-2` |
| 42 U.S.C. § 1320a-7b | `§` | `§` | ShortLawCitation `§ 1320a` (truncated) |
| 42 U.S.C. § 300gg-91 | `§` | `§` | ShortLawCitation `§` |
| 12 U.S.C. § 1749bbb-10c | `§` | `§` | ShortLawCitation `§` |
| 17 CFR § 240.10b-5 | `17 CFR § 240` | `17 CFR § 240` | `17 CFR § 240` |

#331's suffix is uppercase-only (`[A-Z]{0,4}`e lowercase forms don't, even though thechangelog lists `300gg-91` and `1749bbb-10c`. #335's short-cite regex is lowercase-only (`\d+[a-z]?`), so `106A`
falls back to a bare marker and `1320a-7b` cag span. Both PRs point at the real home,reporters-db's `law.section`, which has no letter suffix at all today:

```
(?P<section>(?:\d+(?:[\-.:]\d+){,3})|(?:\d+(?\))+))
```

The `[a-zA-Z]{0,4}` suffix suggested above is the natural first move, but it also swallows prose glued to a number in
OCR'd text (`§1983and`, `§1030is`, `§ 1983th`331's negative tests guard against. U.S. Code letter suffixes are either one letter of either case (`78j`, `2000e`, `1028A`) or a repeated letter (`1395ww`,
`300gg`, `1749bbb`), and glued prose is never the first alternative and spelling therepeats out (no lookahead or backreference, so hyperscan is fine):

```python
rep = "|".join(c * 2 for c in "abcdefghijklmn.join(c * 3 for c in"abcdefghijklmnopqrstuvwxyz")
SFX = f"(?:[A-Za-z]|{rep})"
law_section = rf"(?P<section>(?:\d+{SFX}?(?:[\-.:]\d+{SFX}?){{,3}})|(?:\d+(?:\((?:[a-zA-Z]{{1}}|\d{{1,2}})\))+))"
```

Applied at the same override point #331 uses ractors`:

- every row above becomes a `FullLawCitation` including `240.10b-5` from the originalreport, plus `1395ww`, `1395ww(d)(5)`, `290dd-2`, `1396r-8`, `1437f(o)`, `1395w-4(a)(1)`;
- subsections keep flowing into `pin_cite` (` pin `(a)`; `501(c)(3)` → `501`, `(c)(3)`),which the existing Kan./Ohio/Wis./Ark. law tests require, so I left the second alternative alone;
- `§1983and`, `§1030is`, `§1983or`, `§ 1983th
- `tests/test_FindTest.py` is unchanged against main.

Happy to open this as a PR against reporters-db, or as a patch on top of #331 if @linhtpham09 would rather fold it in, whichever you prefer.

**Comment by mlissner** [comment]
Love that y'all are working on this. Sorry that we don't have more folks on hand to jump in, but hopefully somebody will chime in soon.

**Comment by linhtpham09** [comment]
I am good with either, but I will defer to you @mightymikesapp  since you proposed the solution. I think opening a new PR against reporters-db is good.

**Comment by mightymikesapp** [comment]
Thanks @linhtpham09, and for #331, which got this moving after two years. Opened as freelawproject/reporters-db#279: the `law.section` change above, plus the lettered forms added to the U.S.C. examples and `17 C.F.R. § 240.10b-5` to the C.F.R. examples so the existing example-based tests cover it. It passes reporters-db's suite unchanged, and eyecite main against that branch parses every form in the table above (including the CFR cite from the original report) with the existing test suite unchanged.

Once it lands and releases, eyecite only needs the reporters-db pin bumped and the test cases added. If you'd like to carry that half in #331 so your PR lands as the eyecite side, I'll send you the test matrix; otherwise I'm happy to fold it in. Either way works for me.

**Comment by linhtpham09** [comment]
Feel free to send me the test matrix and I can update the PR

**Comment by mightymikesapp** [comment]
Sent the test matrix to @linhtpham09 on #331.
